File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    private Count compress(Count[] arr) {
        Count result = new Count(0, 0);
        for (Count count: arr) {
            result.zero += count.zero;
            result.one += count.one;
        }
        return result;
    }
    
    private Count compress(int[][] arr, int x, int y, int len) {
        if (len == 1) {
            if (arr[y][x] == 1) {
                return new Count(0, 1);
            } else {
                return new Count(1, 0);
            }
        }
        
        int half = len / 2;
        Count[] compressed = new Count[]{
            compress(arr, x, y, half),
            compress(arr, x + half, y, half),
            compress(arr, x, y + half, half),
            compress(arr, x + half, y + half, half)    
        };
        
        Count result = compress(compressed);
        if (result.zero == 0) return new Count(0, 1);
        if (result.one == 0) return new Count(1, 0);
        return result;
    }
    
    private static class Count {
        int zero;
        int one;
        
        public Count(int zero, int one) {
            this.zero = zero;
            this.one = one;
        }
        
        public int[] toArray() {
            return new int[]{this.zero, this.one};
        }
    }

    
    public int[] solution(int[][] arr) {
        Count count = compress(arr, 0, 0, arr.length);
        return count.toArray();
    }
}

// 최종적으로 남는 0의 개수와 1의 개수를 반환해야 한다.

// 상태 정의
// (x, y, length) -> 해당 정사각형에서 0과 1이 몇개인가

// 종료 조건
// (x, y, length) = [0, 1] -> 1밖에 없거나
// (x, y, length) = [1, 0] -> 0밖에 없음
// (x, y, 1) -> 정사각형 길이가 1임

// 점화식(부분 문제로 현재 문제를 구성), 상태의 전이
// (x, y, length) = 
// (x, y, length / 2) +
// (x + length / 2, y, length / 2) +
// (x, y + length / 2, length / 2) +
// (x + length / 2, y + length / 2, length / 2)

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.kt
class Solution {
    fun solution(arr: Array<IntArray>): IntArray {
        return compress(0, 0, arr.size, arr)
    }
    
    fun compress(x: Int, y: Int, len: Int, arr: Array<IntArray>): IntArray {
        val result = intArrayOf(0, 0)
        if (len == 1) {
            result[arr[y][x]] = 1
            return result
        }
        // 모두 같은지 검사
        var countZero: Int = 0
        var countOne: Int = 0
        for (nextX in x until x + len) {
            for (nextY in y until y + len) {
                if (arr[nextY][nextX] == 0) {
                    countZero++
                } else {
                    countOne++
                }
            }
        }
        if (countZero == 0) {
            return intArrayOf(0, 1)
        } else if (countOne == 0) {
            return intArrayOf(1, 0)
        }
        
        // 분할
        val result1 = compress(x, y, len / 2, arr)
        val result2 = compress(x + len / 2, y, len / 2, arr)
        val result3 = compress(x, y + len / 2, len / 2, arr)
        val result4 = compress(x + len / 2, y + len / 2, len / 2, arr)
        
        result[0] = result1[0] + result2[0] + result3[0] + result4[0]
        result[1] = result1[1] + result2[1] + result3[1] + result4[1]
        
        return result
    }
}